Saved question selection UX

kdoh opened the pull request:
Status - Need help from @daltojohnso to help sort out blockers 

## What this does
Updates the picker you use to select a saved question as a starting point for further exploration to properly account for nested collections and provides more space overall to work.

![image](https://user-images.githubusercontent.com/5248953/108872229-f38d5680-75c7-11eb-8419-9d20e407a73e.png)


How to test

1. Go to "Ask a question" and select "Saved questions"  (see bug below if you only have one database locally)
2. You should see any collections containing saved questions that can be used to start a new question on the left.
3. Clicking on any one of those should show saved questions that are valid starting points for new questions 
4. It should work with large numbers of collections and saved questions

--- 
Project planning scratch (still organizing things here)

Remaining work that has to be done before merge
- [ ] The picker only works for simple mode right now. Not sure how best to also make sure it'll show up in the notebook.
- [ ] Both the collection list and the saved question list probably need to be wrapped in a virtualized list wrapper.
- [ ] Need to test independent scrolling of both the sidebar and list with many items
- [ ] Not sure if we're handling the "no saved questions" case here yet.
- [ ] Handle situation where a parent collection might not have a valid saved q but a child does.
- [ ] There's a "Loading..." state for the entire popover that is quite gross if you hit it (intermittent). We'll need to restructure the markup a bit to have the overall layout stay relatively similar to the final product to prevent flashing.
- [ ] Collection item expand arrows aren't spaced properly

Related issues:
- https://github.com/metabase/metabase/issues/11806 - second bullet point

kdoh committed:
separate out saved questions click trigger

start poking around and start by removing the linkage between saved questions "db" and the whole
accordian list mess. AccordianList assumes there will be one nesting of items in a "schema" as it was
designed to work with databases not our potentially infinitely nestible collections.

instead of conforming to this component we're going to fake it with the initial trigger looking like it's a
part of the same widget but instead it will trigger a different popover to open so we can reflect the structure
users are familiar with.

kdoh committed:
test fit swapping components

this gets a new popover in place when you click Saved Questions (only in simple mode right now).
we keep some basic UI state in the QuestionDataSelector and either show the list of databases or swap over
to our stubbbed our placeholder for saved questions.

kdoh committed:
set up a new component [ci skip]

set up a workspace for further work on the new widget

kdoh committed:
test fit CollectionList [ci skip]

trying out our existing CollectionList in the new SavedQuestionPicker to display collections.
good news is it looks like some of it will work, but we'll need to alter a few things that are
too specific to the initial case. but this seems like a reasonable avenue to continue prototyping with.

kdoh committed:
set up saved question data loading [ci skip]

- filter the collection list based on the saved question schemas endpoint
- load saved questions for a collection

kdoh committed:
file cleanup [ci skip]

kdoh committed:
extract collection list trigger into fn [ci skip]

kind of a gross hack but it lets us have custom rendering for the collection list
so we can keep the same overall pattern but have the action be a little different here

kdoh committed:
[chore] small bit of cleanup [ci skip]

paulrosenzweig committed:
use schemas entity

kdoh committed:
fix styling on saved questions option [ci skip]

kdoh committed:
small styling bits [ci skip]

kdoh committed:
saved question item styling

kdoh committed:
basic sidebar collection selected state

kdoh committed:
lint fixes

mazameli commented:
This isn't working for me. :/ I'm seeing two saved question "databases" and when I click on the second one I'm getting this error: `TypeError: onSwitchToSavedQuestions is not a function. (In 'onSwitchToSavedQuestions()', 'onSwitchToSavedQuestions' is undefined)`

I'm on commit `cfc5547837e09dfb4c28c1a003d483ab398ed89d` and I started with a fresh MB app DB.

![image](https://user-images.githubusercontent.com/2223916/108926265-e5086480-75f2-11eb-92d8-27ca1f1bb528.png)

kdoh commented:
@mazameli Oh I see what's going on. The new picker is only present in simple mode right now. Plan is to get it added in notebook land too if we think the overall thing is viable enough but didn't want to do that extra work just yet.

mazameli commented:
Aaah, gotcha.

mazameli commented:
**Overall**
- Huge improvement. 
- Is search planned for a separate PR/project?
- This could give too _much_ focus to nested queries, and might really only be solving a problem when you have around as many DBs as we do on our Stats instance, but I'm wondering if it might be better to have saved questions be a tab so that you don't have to scroll all the way down past all the DBs. It also kind of separates out your on-disk tables vs. your "custom" tables. Rough/gross sketch:

<img width="566" alt="image" src="https://user-images.githubusercontent.com/2223916/109230179-48240380-7779-11eb-99d1-f7216dcbdd39.png">


**Bug**
A problematic side effect of hiding empty collections is that it also hides any _sub_collections that might be nested underneath the empty parent collection, and the subcollection may have things in it. (Think e.g. of a Releases collection with a subcollection for 0.38, 0.37 etc.)

**Edge case**
My localhost has lots of empty collections, and I didn't immediately understand why I wasn't seeing all of my collections. 

**Out of scope (I'm assuming)**
This is an existent problem on Stats/master, but the flow for going back to the data selection step for Simple mode when using nested query data sucks. For some reason, it thinks that all saved question datasets are part of the sample dataset (which I renamed "core data" for mockup purposes just FYI as you'll see in my gif), so you have to make sure you click Back on your browser rather than using the data breadcrumb:

![going-back-to-pick-data](https://user-images.githubusercontent.com/2223916/109228714-1ad65600-7777-11eb-95d3-e74e7068f4e4.gif)

dpsutton commented:
Small chime in, I've never noticed that there were saved questions in this UI because the icon is so similar
<img width="295" alt="image" src="https://user-images.githubusercontent.com/6377293/109230681-da80d300-778a-11eb-884e-42dc5a5801de.png">

Maybe a better icon, some kind of divider, or some other small visual flag to indicate that this is different?

Commit:
Use virtual list for saved question tables

Commit:
Plug the new saved question picker in the Notebook

This involves:
* moving SavedQuestionPicker into DataSelector component
* use an existing "Saved Question" item, instead of a special one
* propagate schema change (between database and table change)

Commit:
Revert "Use virtual list for saved question tables"

This reverts commit b50338124779716a36fef430d9f72a7616761722.

Commit:
Create an own version of CollectionList for this saved question picker

This also restore the tweaks to customize collections/components/CollectionsList.

Commit:
Sync the step (Database->Schema->Table) with the saved question picker

Conditionally render the saved question picker as soon as it's evidently
clear that this is that special database. That means, no manual
databaset set or schema change is triggered.

ariya commented:
Because of a radically different approach to hook this into the notebook, to make it easier to follow the changes, I reconstructed this PR as #15663.
